Enables running transitions synchronously

dianabarsan opened the pull request:
# Description

Pulls sentinel transitions out into a shared library
Enables usage of new shared library @medic/transitions in Sentinel and API.
API `sms` and `records` endpoints now run transitions over docs before saving.

#4857


# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

dianabarsan commented on shared-libs/transitions/src/schedule/due_tasks.js:
```diff
@@ -51,9 +50,15 @@ module.exports = {
+          objs[row.id].keys.push(row.key);
+
+          return objs;
+        }, {});
```
Fix for updating all valid `scheduled_tasks` instead of one.

garethbowen replied:
I know this is a pain but can we separate this out as a separate fix? It'd make it much easier to find later. It can wait until this PR is merged.

dianabarsan replied:
Sure.

dianabarsan replied:
Changes have been reverted.

dianabarsan commented on shared-libs/transitions/src/transitions/index.js:
```diff
@@ -161,8 +136,9 @@ const loadTransitions = () => {
+    const allowSync = conf && !conf.async;
 
-    if (!conf || disabled) {
+    if (!conf || disabled || (!allowSync && synchronous)) {
```
I chose to make the config "disable" some transitions from running synchronously instead of "enabling" to run synchronously (after back and fourth-ing a few times). Reasoning was to enable projects to start using this feature without changes to configurations.
AFAICT the only transitions which **should** run async are `muting` and `update_notifications`, which may end up needing to updated lots of documents.

I would appreciate some feedback on this.

garethbowen replied:
I'm wondering if we should just hardcode those ones as async and not allow configuration override. Something like...

```js
module.exports = {
  filter: () => {...},
  onMatch: () => {...},
  asynchronousOnly: true // defaults to false
};
```

We definitely want to default to sync for the rest of the transitions. The most important bits to sync are the validations and replies.

dianabarsan replied:
Ok, great! Will do.

garethbowen replied:
I'm not sure I agree that "any transition that is sync safe can be run in parallel / in batches", but maybe I don't understand what you mean...

> I'm pretty sure that multi_report_alerts need to be only run async, because it counts reports written to disk, and so if done in a batch you could miss some.

I don't think sync exec has anything to do with batching. I think it's fine to run it sync but not batched. The only downside is it's quite a slow transition so it'll slow down the response massively. It's also not that useful to do in sync because it usually doesn't respond to the sender but some supervisor.

dianabarsan replied:
Updated so transitions are forced to by async only by hardcoded `asynchronousOnly`.
`muting`, `update_notifications` and `multi_report_alerts` run async only.

SCdF replied:
> I'm not sure I agree that "any transition that is sync safe can be run in parallel / in batches"

We can wax on at standup about what this kind of things means, it's kind of hard to do in PR comments. It's also kind of confusing because sync and async are overloaded here to mean potentially opposite things (e.g. when I say sync do I mean sync as part of SMS writing, or sync in that they run one after the other in sentinel, and visa versa for async).

What I meant there though, is that this is a true and required property of running a transition before writing the SMS by definition of this PR (unless I misread it, in which case I apologise!, or it changes), because in the PR SMS **are** being run through in a batch, so it has to be part of the decision making on which transitions are OK to run here[1]. I can see your additional point about performance (not bothering to run transitions sync that aren't going to potentially generate responses), which is a great thought here as well which I did not consider.

[1] e.g. you have a multi-report-alert configured to send an SMS if you see three malaria reports in 24 hours, there is already one on disk, you get two in the batch of SMS from gateway: if this transition ran at SMS-write then both reports would think there are two malaria reports, as opposed to the first thinking there is two and the second thinking there is three.

dianabarsan replied:
Hmm, I think this is indeed an issue with other transitions as well. 
Fe: `accept_patient_reports` with clearing schedules. 

Will have to update so docs are saved after transitions run over them, in sequence versus in batch.

dianabarsan commented on shared-libs/transitions/src/transitions/index.js:
```diff
@@ -285,10 +246,15 @@ const finalize = ({ change, results }, callback) => {
+    );
+    return callback();
+  }
+
```
I moved the call to `canRun` from the `applyTransitions` map function.
The reason was to test `canRun` over the document after it has been updated by previous transitions.

As an example, it enables the possibility of running `registration` over a fresh doc: currently `registration` requires either the form to be public or `doc.contact` to exist, while `update_clinics` is responsible for creating `doc.contact`.

SCdF replied:
This is a good idea.

We should also formalise transition run order to make this less fragile.

dianabarsan replied:
We already formalize that, since 2.15
https://github.com/medic/medic/blame/d7ec779eac1806a26cd6f546970b9319648a3cc7/shared-libs/transitions/src/transitions/index.js#L13

dianabarsan commented on shared-libs/transitions/src/transitions/index.js:
```diff
@@ -96,60 +52,79 @@ const processChange = (change, callback) => {
         });
+        async.series(operations, (err, results) => {
+          return err ? reject(err) : resolve(results);
+        });
```
Just thought about this.. do you guys think I should delete the just-created info-docs for docs that are just echo-ed back?

garethbowen replied:
I think this is fine for now - deleting a doc is probably worse for performance than just leaving it there forever...

I think eventually we want the transitions function to return an array of docs to save, so all the docs are saved once all transitions have executed, but that would need careful consideration...

dianabarsan replied:
I would very much agree with the transaction-like approach.

dianabarsan replied:
If the transition returns "not changed" or has an error, I return the untouched version of the document that I received, which will not even have an `_id` property. 
The reason is that some transitions add errors on the docs, but still return as "not changed". 
The doc will receive another `_id` upon save, so the info-doc we just created is useless. 
I guess I could only copy the `_id` property over in this case, so it wouldn't be useless?

SCdF replied:
Might as well right? The upside I presume being that when these docs are written Sentinel won't re-run those transactions over them. And the main downside is that we don't generate "smart" UUIDs that CouchDB likes, but honestly we don't do that for 90% of our data right now anyway (cause of PouchDB), and if we really wanted to we could replace the uuid generation used here with a cache that [sources UUIDs from CouchDB](https://docs.couchdb.org/en/stable/api/server/common.html#get--_uuids) if we considered it an issue.

dianabarsan replied:
Sentinel would run all transitions over that doc either way.

SCdF replied:
That's a shame. Maybe not in the future though (like I feel like we _should_ be able to take advantage of this even if we aren't now)!

Also, it's probably better for Impact to not have orphaned info docs. I may not matter depending on their impl, but it's a bug waiting to happen in any case.

dianabarsan replied:
This thread is outdated because docs are saved in series within the library immediately after running transitions, same as old sentinel.

dianabarsan commented:
I guess this is it.. new and improved along with as much git history. 
The branch itself now uses the shared lib published under another name. 

Could you please have a look, @garethbowen & @SCdF . Thank you!

garethbowen commented on shared-libs/transitions/src/config.js:
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+  init: (c, t) => {
+    config = Object.assign({}, c);
+    translations = Object.assign({}, t);
```
Do you need to shallow clone the translations? Will this still be updated if the translations are modified?

dianabarsan replied:
I'm calling `init` in both API and Sentinel whenever translations or updated. 
-> https://github.com/medic/medic/blob/4857-detached-sentinel-bis/api/src/config.js#L173
-> https://github.com/medic/medic/blob/4857-detached-sentinel-bis/sentinel/src/config.js#L48

I can remove the shallow clone.

garethbowen replied:
Ahh I must've missed that. Sorry.

garethbowen commented on shared-libs/transitions/src/db.js:
```diff
@@ -0,0 +1,39 @@
+
+module.exports.init = db => {
+  Object.assign(module.exports, _.omit(db, 'init'));
+};
```
Is this solely for testing? If so, can we move it into the tests dir for clarity?

dianabarsan replied:
Well, not for testing. 
https://github.com/medic/medic/blob/4857-detached-sentinel-bis/shared-libs/transitions/src/index.js#L5 

I'm passing the whole `db` and whole `logger` objects to this library so it simulates functioning in the environment that it was called from. 

The DB-s, for example, have headers attached to the `fetch` functions (`X-Medic-Service: api`, `X-Medic-User: sentinel`) for haproxy auditing. 
I thought about just passing these specific headers and duplicate the code to initialize the `db` but went with this instead. 
I'll update them to just pass the headers and other specific info needed.

garethbowen replied:
Why not just use the sourceDb directly instead of shallow cloning it?

garethbowen commented on shared-libs/transitions/src/lib/logger.js:
```diff
@@ -0,0 +1,15 @@
+
+module.exports.init = logger => {
+  Object.assign(module.exports, _.omit(logger, 'init'));
+};
```
Same here - I'm assuming it's testing only?

garethbowen commented on tests/e2e/registration-by-sms.js:
```diff
@@ -262,11 +262,11 @@ describe('registration transition', () => {
       const taskElement = element(by.css('#reports-content .details > ul'));
       expect(taskElement.element(by.css('.task-list > li:nth-child(1) > ul > li')).getText()).toBe('Thank you '+ CAROL.name +' for registering Siobhan');
-      expect(taskElement.element(by.css('.task-list > li:nth-child(1) .task-state .state.pending')).isDisplayed()).toBeTruthy();
+      expect(taskElement.element(by.css('.task-list > li:nth-child(1) .task-state .state.forwarded-to-gateway')).isDisplayed()).toBeTruthy();
```
WOOHOO!

garethbowen commented on tests/e2e/sentinel/transitions/generate_patient_id_on_people.spec.js:
```diff
@@ -2,7 +2,7 @@ const utils = require('../../../utils'),
       uuid = require('uuid');
 
-describe('death_reporting', () => {
+describe('generate_patient_id_on_people', () => {
```
Hah

SCdF commented on shared-libs/transitions/src/schedule/due_tasks.js:
```diff
@@ -1,7 +1,6 @@
 // TODO: this doesn't need to exist. We can just work this out dynamically when
 //       gateway queries, it's not slow or complicated.
 const async = require('async'),
-  _ = require('underscore'),
```
What's the reasoning behind moving the schedule code into the transitions shared lib? AFAICT it's only used by sentinel.

dianabarsan replied:
It's funny you should ask! 
It's because _someone_ wrote this two years ago: 
https://github.com/medic/medic/blame/8c25c15e6fec268b272d5fb9b361bd119864a4e6/sentinel/src/schedule/due_tasks.js#L1

garethbowen replied:
@dianabarsan Could you please raise an issue so this actually gets fixed? I'm happy to do it but you're the sentinel expert now!

dianabarsan replied:
sure!

dianabarsan commented on shared-libs/transitions/src/transitions/index.js:
```diff
@@ -96,60 +52,79 @@ const processChange = (change, callback) => {
+// - hydrate the docs and generate info docs
+// - run loaded transitions over all docs
+// - returns docs, either updated or their original version
+const processDocs = docs => {
```
I guess we could, but then everything would be done one doc at a time:
 - fetching lineage
 - creating the infodoc 

Perf wise, it would be much worse than bulking them.

garethbowen commented on shared-libs/transitions/src/transitions/index.js:
```diff
@@ -96,60 +52,79 @@ const processChange = (change, callback) => {
-            return;
-          }
+  let changes;
+  const clonedDocs = JSON.parse(JSON.stringify(docs)),
```
Yeah every time I check this is what it comes back to. Last time I checked was a couple of months ago.

dianabarsan replied:
I guess it depends on what you're cloning. 
As a result of Stefan's comment, I tried a couple of _reliable_ alternatives : http://jsben.ch/SJghM
- [lodash cloneDeep](https://lodash.com/docs/4.17.11#cloneDeep) 
- [angular copy](https://docs.angularjs.org/api/ng/function/angular.copy)

Do you guys think lodash.cloneDeep would be worth including? It can be required as a standalone.

garethbowen replied:
In my Chromium, JSON is faster than lodash but angular is way faster. Do you get lodash as fastest?

![screenshot from 2019-03-08 09-32-34](https://user-images.githubusercontent.com/2064938/53987176-2349b680-4185-11e9-9556-33aa76dcbb42.png)

dianabarsan replied:
_slightly_ faster

![image](https://user-images.githubusercontent.com/35681649/53987481-ca006980-4129-11e9-9431-e770d29d1cb2.png)

garethbowen replied:
Mine's consistently slower. Strange. Different browser versions maybe?

I wonder how angular copy is doing it?

I don't think it's worth switching to lodash but it may be worth ripping out the angular version and using that throughout in a separate PR.

dianabarsan replied:
I was planning on looking at the Angular version tomorrow, it got me curious. 

The lodash seems to be the complete solution to deep copy and tries to treat all possible cases, while angular's solution is clearly stated to be riddled with caveats, albeit faster.

dianabarsan commented:
I've rebased to resolve a conflict. Also published the shared-lib under the correct name and updated the dependencies in the apps. Had to do some minor updates to get a couple of unit tests passing. 
All changes included in this [commit](https://github.com/medic/medic/pull/5476/commits/95e90993ff228a6c378e46c303611d79d0621ff2)
